src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -216,14 +251,29 @@ public void OnResponseHeadersStart()
                 }
             }
 
+            public void OnResponse100Continue()
+            {
+                // This is called when we receive complete set of headers with status 100 Continue.
+                // We need to reset state and wait for another final response.
+                lock (SyncObject)
+                {
+                    if (_state != StreamState.ExpectingData)
```

Comment:
It seems weird that we first transition to ExpectingData, then transition *back* to ExpectingHeaders here. Isn't there a timing issue?

Seems like we should check for 1xx while we've got the lock when we are processing headers (in TryEnsureHeaders, I think) and just not transition to ExpectingData if we get a 1xx.

Reply:
Actually, TryEnsureHeaders is the wrong place, ignore that. I'll figure out what the right place is.

Reply:
The right place to check for 1xx response codes is in `OnResponseHeadersComplete`, I think.

We should check for 1xx status code there under the lock, and if so:
(1) Leave the state as ExpectingHeaders. (Assuming it was already ExpectingHeaders, not ExpectingTrailingHeaders.)
(2) If we have a 100 continue task, complete it and clear it out so it won't be completed again.

We should not need the OnResponse100Continue method at all.

Reply:
And there's more we need to do, actually. We need to ignore any headers that come in on a 1xx informational response -- these should not be added to the response object's headers collection. That's how we handle these for HTTP/1.1.

This means we actually need to detect and remember that we saw the 1xx status. This must always be the first header, so it should be relatively straightforward to do.

There's a related issue here (tracked somewhere, I can dig it up if that helps): Status code must always be the first header in the header block. That's good, because it means we can detect 1xx before processing other headers. We don't enforce this today, however.

We should probably split the ExpectingHeaders state into a couple different states:

(1) ExpectingStatus. In this state, we fail if we receive anything other than a status code. If we do, we check for 1xx and transition to one of the states below.
(2) ExpectingHeaders. We received a final response code, so process headers and add to response object.
(3) ExpectingIgnoredHeaders. We received 1xx, so process headers and ignore them.

Then in OnResponseHeadersComplete, ExpectingHeaders will transition to ExpectingData, while ExpectingIgnoredHeaders will transition back to ExpectingStatus.

Reply:
There doesn't seem to be a github issue for the status code issue I mentioned above. Regardless, this is something we should address.

Reply:
Ah, I see that we are enforcing that status code is always the first header. We do this by checking whether _response is null.

So all we need to do here, I think, is add a status like ExpectingIgnoredHeaders (or whatever name). When we read the status code, if it's 1xx then change the state to ExpectingIgnoredHeaders, and don't create the _response object. When we read a regular header, just ignore it if the state is ExpectingIgnoredHeaders.

Reply:
yes. I'll take a look. that should be easy.